src/main/java/picard/sam/SamToFastq.java

```diff
@@ -466,40 +474,51 @@ private void assertPairedMates(final SAMRecord record1, final SAMRecord record2)
      * Allows for lazy construction of the second-of-pair writer, since when we are in the "output per read group mode", we only wish to
      * generate a second-of-pair fastq if we encounter a second-of-pair read.
      */
-    static final class FastqWriters {
+    private static final class FastqWriters {
         private final FastqWriter firstOfPair, unpaired;
         private final Lazy<FastqWriter> secondOfPair;
+        private final Map<SAMReadGroupRecord, List<FastqWriter>> additionalWriters;
 
-        /** Constructor if the consumer wishes for the second-of-pair writer to be built on-the-fly. */
-        private FastqWriters(final FastqWriter firstOfPair, final Lazy<FastqWriter> secondOfPair, final FastqWriter unpaired) {
+        /**
+         * Constructor if the consumer wishes for the second-of-pair writer to be built on-the-fly.
+         */
+        private FastqWriters(final FastqWriter firstOfPair, final Lazy<FastqWriter> secondOfPair, final FastqWriter unpaired,
+                             Map<SAMReadGroupRecord, List<FastqWriter>> additionalWriters) {
             this.firstOfPair = firstOfPair;
             this.unpaired = unpaired;
             this.secondOfPair = secondOfPair;
+            this.additionalWriters = additionalWriters;
         }
 
-        /** Simple constructor; all writers are pre-initialized.. */
-        private FastqWriters(final FastqWriter firstOfPair, final FastqWriter secondOfPair, final FastqWriter unpaired) {
-            this(firstOfPair, new Lazy<>(() -> secondOfPair), unpaired);
+        /**
+         * Simple constructor; all writers are pre-initialized..
+         */
+        private FastqWriters(final FastqWriter firstOfPair, final FastqWriter secondOfPair, final FastqWriter unpaired,
+                             Map<SAMReadGroupRecord, List<FastqWriter>> additionalWriters) {
+            this(firstOfPair, new Lazy<>(() -> secondOfPair), unpaired, additionalWriters);
         }
 
-        public FastqWriter getFirstOfPair() {
+        private FastqWriter getFirstOfPair() {
             return firstOfPair;
         }
 
-        public FastqWriter getSecondOfPair() {
+        private FastqWriter getSecondOfPair() {
             return secondOfPair.get();
         }
 
-        public FastqWriter getUnpaired() {
+        private FastqWriter getUnpaired() {
             return unpaired;
         }
 
-        public void closeAll() {
-            final Set<FastqWriter> fastqWriters = new HashSet<FastqWriter>();
+        private void closeAll() {
+            final Set<FastqWriter> fastqWriters = new HashSet<>();
             fastqWriters.add(firstOfPair);
             fastqWriters.add(unpaired);
+            fastqWriters.addAll(additionalWriters.values().stream().flatMap(Collection::stream).collect(Collectors.toList()));

```

Comment:
I think `additionalWriters.values().forEach(fastqWriters::addAll);` is equivalent